### PR TITLE
 VSCode : Enable Development inside Docker Container

### PR DESCRIPTION
**Pull request by embetrix**
This will enable seamless Development of optee_client Project with VSCode
inside Docker container.

The VSCode Container provides:

* Automatic Mapping of the host user UID/GID inside the container
* Cross-Compilers (aarch64/armhf)
* GDB Multiarch for Debugging 
* Remote SSH on the Target with seamless ssh mapping of the Host inside the container (Linux/Windows WSL)
* Git and Git Extensions to work and send Patches

Note: This require Docker installation for Linux/or Windows :
      [1] https://docs.docker.com/engine/install

      VSCode Extensions (extensions.json) will be recommended for the installation

For more Information :
[2] https://code.visualstudio.com/docs/devcontainers/containers
[3] https://www.youtube.com/watch?v=C_5tDWsWSj0
[4] https://www.youtube.com/watch?v=b1RavPr_878

### CODE REVIEW

**Comment by jforissier** [comment]
Hello @embetrix,

That looks useful, but could you give more information in the commit description? Such as how this is typically used in a development environment. I do not ask for a very detailed documentation but a few commands and explanations on the steps that are involved to build and run the project. Thanks!

**Comment by embetrix** [comment]
> Hello @embetrix,
> 
> That looks useful, but could you give more information in the commit description? Such as how this is typically used in a development environment. I do not ask for a very detailed documentation but a few commands and explanations on the steps that are involved to build and run the project. Thanks!

@jforissier : I added some information, normally everthing should be automatically set by VSCode just by accepting the extension recommendations and openning the project inside the conatiner

**Comment by embetrix** [comment]
@jforissier any update ?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+ENV LANG=C.UTF-8
+
+RUN dpkg --add-architecture armhf
+RUN dpkg --add-architecture arm64
+
+RUN echo 'deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ jammy main restricted universe multiverse' > /etc/apt/sources.list
+RUN echo 'deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ jammy-updates main restricted universe multiverse' >> /etc/apt/sources.list
+RUN echo 'deb [arch=amd64] http://archive.ubuntu.com/ubuntu/ jammy-backports main restricted universe multiverse' >> /etc/apt/sources.list
+RUN echo 'deb [arch=amd64] http://security.ubuntu.com/ubuntu/ jammy-security main restricted universe multiverse' >> /etc/apt/sources.list
```
Comment (jforissier) [inline Dockerfile:11]: Why is this needed? Isn't this in the ubuntu:22.04 image already?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    automake \
+    cmake \
+    gcc-aarch64-linux-gnu \
+    gcc-arm-linux-gnueabihf \
+    gdb-multiarch \
+    curl \
+    dpkg-dev \
+    pkg-config \
+    uuid-dev:armhf \
+    uuid-dev:arm64
```
Comment (jforissier) [inline Dockerfile:29]: In https://optee.readthedocs.io/en/latest/building/prerequisites.html we have `uuid-dev` (host architecture). Why are `:armhf` and `:arm64` needed?

Comment (embetrix) [reply Dockerfile:29]: I took all those from CI Dockerfile: https://github.com/OP-TEE/optee_client/blob/master/ci/Dockerfile.ubuntu, I just extended it to include additional dev tools.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{
+	"name": "optee-client-docker",
+	"build": {
+		"dockerfile": "../Dockerfile"
```
Comment (jforissier) [inline .devcontainer/devcontainer.json:4]: Could we rename this file to `Dockerfile.vscode` and store it here under `.devcontainer/`?

File: Dockerfile (hunk 2 above)
Comment (jforissier) [reply Dockerfile:29]: OK now I remember, they are required for the CMake build which uses `pkg-config` for cross-compilation.

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (embetrix) [reply .devcontainer/devcontainer.json:4]: sure

Comment (embetrix) [reply .devcontainer/devcontainer.json]: done

**Review by jforissier — APPROVED** [review]
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Thanks!

File: .devcontainer/Dockerfile.vscode [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    automake \
+    cmake \
+    gcc-aarch64-linux-gnu \
+    gcc-arm-linux-gnueabihf \
+    gdb-multiarch \
+    curl \
+    dpkg-dev \
+    pkg-config \
+    uuid-dev:armhf \
+    uuid-dev:arm64
```
Comment (jforissier) [inline .devcontainer/Dockerfile.vscode:29]: Would this work instead:
```
# syntax = edrevo/dockerfile-plus

INCLUDE+ ../ci/Dockerfile.ubuntu
```

I would like to avoid unnecessary duplication.

**Review by jforissier — COMMENTED** [review]
Please see comment below. Thanks.

File: .devcontainer/Dockerfile.vscode (hunk 4 above)
Comment (embetrix) [reply .devcontainer/Dockerfile.vscode:29]: @jforissier i checked this a while ago, vscode does not support it :-(

Comment (jforissier) [reply .devcontainer/Dockerfile.vscode:29]: Ah, OK. Thank you. Then please add my Acked-by and I will merge this.

Comment (embetrix) [reply .devcontainer/Dockerfile.vscode:29]: done
